File: BFS/rian00/5014.js
const { sep } = require("path");

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

class Info {
  constructor(_pos, _step) {
    this.pos = _pos;
    this.step = _step;
  }
}
const [F, S, G, U, D] = input[0].split(" ").map((e) => +e);

const visited = new Array(F + 10).fill(false);

const deque = [];
deque.push(new Info(S, 0));
visited[S] = true;
let flag = false;

const di = [U, -D];

while (deque.length > 0) {
  q = deque.shift();
  if (q.pos === G) {
    console.log(q.step);
    flag = true;
    break;
  }

  for (let i = 0; i < 2; i++) {
    const newPos = q.pos + di[i];
    if (newPos >= 1 && newPos <= F && !visited[newPos]) {
      deque.push(new Info(newPos, q.step + 1));
      visited[newPos] = true;
    }
  }
}

if (!flag) {
  console.log("use the stairs");
}

File: bfs/JunYoungKr/5014.py
from collections import deque
import sys

input = sys.stdin.readline

MAX = 1000001
visited = [0] * MAX

def bfs(S):
    queue = deque([S])
    
    # 출발지는 방문했으므로 1로 변경
    visited[S] = 1
    
    while queue:
        v = queue.popleft()
        
        # 목적지에 도착한 경우
        if v == G:
            return visited[v] - 1
        
        for i in (v + U, v - D):
            if 0 < i <= F and visited[i] == 0 and i != S:
                visited[i] = visited[v] + 1
                queue.append(i)
    return "use the stairs"

F, S, G, U, D = map(int, input().split())
# S : 출발지
# G : 목적지
# U : 위로 U층으로 가는 버튼
# D : 아래로 D층을 가는 버튼

# U층 위, 또는 D층 아래에 해당하는 층이 없을 때는, 엘리베이터는 움직이지 않는다
print(bfs(S))    